src/Lecture2.hs

```diff
@@ -159,8 +180,43 @@ data Knight = Knight
     , knightAttack    :: Int
     , knightEndurance :: Int
     }
-
-dragonFight = error "TODO"
+data Color
+  = Red
+  | Green
+  | Black
+
+
+data Treasure
+  = Gold
+  | GoldPlusTreasure
+
+data Dragon = Dragon
+  { dragonColor :: Color
+  , dragonExperiencePoints :: Int
+  , dragonTreasure :: Treasure
+  , dragonFirePower :: Int
+  , dragonHealth :: Int
+  }
```

Comment:
That's a great idea! And almost the perfect solution 🔝 

There're two problems with this particular data type definition:

1. It's possible to have a treasure inside green dragon:

```haskell
myDragon = Dragon
    { dragonColor = Green
    , dragonExperiencePoints = 50
    , dragonTreasure = GoldPlusTreasure
    , dragonFirePower = 70
    , dragonHealth = 150
    }
```

2. It's possible to not have a treasure inside non-greed dragons:

```haskell
myDragon = Dragon
    { dragonColor = Red
    , dragonExperiencePoints = 50
    , dragonTreasure = Gold
    , dragonFirePower = 70
    , dragonHealth = 150
    }
```

Reply:
@chshersh You are right, I know I cut the corners a bit on this one :-) I enjoyed this particular exercise a lot as it was like a fantasy story with dragons, knights and treasures !! :-) I got the point about making invalid states unrepresentable, but could not make much progress towards it. I started watching the video you suggested, but wasn't able to watch it completely yesterday. I'm planning to go over it again.